src/algebra/squares.lean

```diff
@@ -0,0 +1,262 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.algebra.basic
+import linear_algebra.finsupp
+import algebra.algebra.tower
+
+/-!
+#  Squares
+
+Let `R` be a Type with `has_mul R`.
+
+We introduce the predicate `is_square` on elements `r ∈ R`, asserting existence of an element
+`s ∈ R`, such that the equality `r = s * s` holds.
+
+Progressively assuming more of the type `R` we
+* introduce the subtype `square R` of squares in `R`;
+* introduce sums of squares, when `R` is an additive commutative monoid;
+* prove an instance of `comm_semigroup` for `squares R`, if `R` is a `comm_semigroup`;
+* prove an instance of `comm_monoid` for `squares R`, if `R` is a `comm_monoid`;
+* prove an instance of `comm_semiring` for `sums_of_square R`, if `R` is a `comm_semiring`.
+-/
+variables {R : Type*}
+
+section is_square
+
+section has_mul
+
+variables (R) [has_mul R]
+
+/--  The predicate `is_square` on `r ∈ R` is the assertion that there exists an element
+`s ∈ R` such that `r = s * s`. -/
+def is_square (r : R) : Prop :=
+∃ s : R, r = s * s
+
+/--  The squares of a Type `R` with `has_mul` is the subtype of the elements of the form `s * s`,
+for some `s ∈ R`. -/
+def squares : Type* := subtype (is_square R)
+
+/--  The sums of squares of an `add_comm_monoid R` is the `ℕ`-submodule spanned by the squares. -/
+def sums_of_squares [add_comm_monoid R] : submodule ℕ R :=
+submodule.span ℕ { r : R | is_square R r }
```

Comment:
Any reason for using the `submodule` and not the `sub_monoid`?

Reply:
Just that I did not know about submonoid.  I can change it to submonoid, if that is preferred!

Reply:
Now that I have tried to make the conversion, a better reason seems to be that I know the library of `submodule` better than the one for `add_submonoid` and, in particular, I could not find a statement like `submodule.mem_span` for `add_submonoid`.

Reply:
Are you looking for `add_submonoid.closure`?

Reply:
Actually, that I had found and now I also fixed all remaining proofs.  I am going to push the result!